cmd/charmload/main.go

```
@@ -0,0 +1,263 @@
+	// Retrieve the branch with a lightweight checkout, so that it
+	// builds a working tree as cheaply as possible. History
+	// doesn't matter here.
+	tempDir, err := ioutil.TempDir("", "publish-branch-")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+	branchDir := filepath.Join(tempDir, "branch")
+	logger.Debugf("running bzr checkout ... %v", branchURL)
+	output, err := exec.Command("bzr", "checkout", "--lightweight", branchURL, branchDir).CombinedOutput()
+	if err != nil {
+		return outputErr(output, err)
+	}
+
+	tipDigest, err := bzrRevisionId(branchDir)
+	if err != nil {
+		return err
+	}
+	if tipDigest != digest {
+		digest = tipDigest
+		logger.Warningf("tipDigest %v != digest %v", digest, tipDigest)
+	}
+
+	logger.Tracef("read CharmDir from branchDir %v", branchDir)
+	thisCharm, err := charm.ReadCharmDir(branchDir)
+	if err != nil {
+		return err
+	}
+	reader, writer := io.Pipe()
+	hash1 := sha512.New384()
+	var counter Counter
+	mwriter := io.MultiWriter(hash1, &counter)
+	thisCharm.ArchiveTo(mwriter)
+	hash1str := fmt.Sprintf("%x", hash1.Sum(nil))
+	go func() {
+		thisCharm.ArchiveTo(writer)
+		writer.Close()
+	}()
+	id := URLs[0]
+	URL := storeURL + id.Path() + "/archive?hash=" + hash1str
+	logger.Infof("posting to %v", URL)
+	request, err := http.NewRequest("POST", URL, reader)
+	if err != nil {
+		return err
+	}
+	authhash := base64.StdEncoding.EncodeToString([]byte(storeUser))
+	logger.Tracef("encoded Authorization %v", authhash)
+	request.Header["Authorization"] = []string{"Basic " + authhash}
+	// go1.2.1 has a bug requiring Content-Type to be sent
+	// since we are posting to a go server which may be running on
+	// 1.2.1, we should send this header
+	// https://code.google.com/p/go/source/detail?r=a768c0592b88
+	request.Header["Content-Type"] = []string{"application/octet-stream"}
+	request.ContentLength = int64(counter)
+	resp, err := http.DefaultClient.Do(request)
```

IMO, its overkill for an error that I've only observed when using incorrect credentials and I'd have to do more to bubble up the fail fast UnauthorizedError.